### DISCUSSION TITLE
Is it possible to have a custom thread pool to execute GstCallback (AppSink.connect) ? (question)

### DISCUSSION THREAD

**Discussion by lmcdasi** [gstreamer-java/gst1-java-core#251] [Q&A] [answered]
Hi,

  I would like to know if it's possible to have a custom thread pool servicing the GstCallback's instead of using the GST default one ?

  I've seen a GstTaskPool - but I did not found any example on how to do so in java.

Thx.

**Comment by neilcsmith-net** [comment]
At present, no (there is some support for main context).  What's your use case exactly?  It has crossed my mind before but not really had a compelling enough need.

It should also never be a thread pool as such - needs to be single threaded.

Please keep questions in discussions or on the mailing list, thanks.

**Comment by lmcdasi** [comment]
I'm trying to improve the performance of an app and my gstream knowledge is not that great so I may err.
When the pipeline is set, I do have:
appSink.connect((AppSink.NEW_SAMPLE) (AppSink appSink) -> {.. process incoming data ..})

When I stream voice media, each time I have a new buffer I will get a Gst thread that executes {.. process incoming data ..} and I was wondering if I can have that processing data through a thread pool using a blocking queue to preserve the sequence. In this way the Gst thread will be released faster rather then blocking until the process is done on that buffer. On that thread pool I have can defined the no. of threads that I want to.

At the same time, for each media I'm having a new pipeline which looks to be 'heavy'. I'm using udpsrc and I noticed that the rtsp uses an rtp demux so I'm wondering if I can create only one pipeline listening on only one port & with rtpdemux/SSRC I can split the media in separate let's ay files.

Thx.

**Comment by neilcsmith-net** [comment, marked as answer]
In which case, the thread pool in `Gst` isn't relevant. Buffer callbacks happen on the native streaming thread. You can do what you like with the `Sample`, including passing it to other threads to process, just make sure not to `dispose()` it until you're done.  GStreamer uses a buffer pool, so be careful about memory use here - the longer you keep the samples / buffers, the more will be allocated by GStreamer.

**Reply by lmcdasi** [reply]
Yes. I have one more question. My bottleneck is not memory, it is CPU usage. The rate of the incoming pkts is 160 bytes every 20 ms.
Thus I get a Gst thread with the sample to process at that rate. 
Is it possible to reduce the sample rate and have gstreamer 'merge' a certain samples into one before executing the callback ?
At that rate on the vm node (only 2 cpu's) that I have I'm able to reach only 300 parallel streams & I need to go higher ...
Thx.

**Reply by neilcsmith-net** [reply]
Probably. Without knowing your pipeline(s), or what you're doing with the data (processing in JVM or elsewhere), it's hard to say what the answer is - whether to batch up data in GStreamer or on the Java side.  But all the samples from any one pipeline should usually be arriving on the same (GStreamer allocated) thread.  Can't really help more than that without deep diving into the architecture and configuration of your project, which is a little more than I can offer via free support right now.

**Reply by lmcdasi** [reply]
Thank you.

**Reply by lmcdasi** [reply]
If by any chance you can share a tip/link to a sample on how I could buffer more data in gstreamer before getting the sample it will be much appreciated. Thank you.

**Reply by neilcsmith-net** [reply]
There isn't _a_ tip for this.  I suggest you ask upstream at https://lists.freedesktop.org/mailman/listinfo/gstreamer-devel explaining what exactly you want to do and sharing the pipeline description(s) you have at present.

**Reply by neilcsmith-net** [reply]
@lmcdasi marking this as answered, as this covers the basic question, and to the best of my knowledge you're no longer using GStreamer and GStreamer-Java bindings for the specific project.
